hard/ultrascale_plus/bsg_link/tcl/bsg_link_ddr.sample_constraints.tcl

```diff
@@ -0,0 +1,40 @@
+# User should re-define periods / names / ports / pins / margins
+# Then add the following lines into the .xdc constraint file
+
+# create input clock
+set fmc_input_clk_period         2.000
+set fmc_input_clk_name           fmc_clk_in
+set fmc_input_clk_port           [get_ports fmc_clk_i]
+create_clock -name $fmc_input_clk_name -period $fmc_input_clk_period $fmc_input_clk_port
+
+# create output clock
+set fmc_output_clk_name          fmc_clk_out
+set fmc_output_clk_pin           [get_pins uplink/ch[0].oddr_phy/ODDRE1_clk/C]
+set fmc_output_clk_port          [get_ports fmc_clk_o]
+create_generated_clock -name $fmc_output_clk_name -source $fmc_output_clk_pin -edges {1 2 3} -edge_shift {0 0 0} $fmc_output_clk_port
+
+# input delay margins
+set dv_bre                 1.0
+set dv_are                 1.0
+set dv_bfe                 1.0
+set dv_afe                 1.0
```

Comment:
This is the constraint template from Vivado for Center-Aligned DDR source synch inputs:
```# Center-Aligned Double Data Rate Source Synchronous Inputs
#
# For a center-aligned Source Synchronous interface, the clock
# transition is aligned with the center of the data valid window.
# The same clock edge is used for launching and capturing the
# data. The constraints below rely on the default timing
# analysis (setup = 1/2 cycle, hold = 0 cycle).
#
#                       |<--------------- 2.5ns ----------------->|
# input                  ____________________
# clock    _____________|                    |____________________|
#                       |                    |
#                dv_bre | dv_are      dv_bfe | dv_afe
#               <------>|<------>    <------>|<------>
#          _    ________|________    ________|________    _
# data     _XXXX____Rise_Data____XXXX____Fall_Data____XXXX_
#
# Reference design VU47P had 500MHz and set these values to 1, which is unrealistic and gives a 2ns capture
# window and doesn't model any degradation in the system...
# For this design, at 400MHz, the period is 2.5, so changing to 1.25

# set input_clock         <clock_name>;      # Name of input clock
# set input_clock_period  <period_value>;    # Period of input clock (full-period)
# set dv_bre              0.000;             # Data valid before the rising clock edge
# set dv_are              0.000;             # Data valid after the rising clock edge
# set dv_bfe              0.000;             # Data valid before the falling clock edge
# set dv_afe              0.000;             # Data valid after the falling clock edge
# set input_ports         <input_ports>;     # List of input ports

# # Input Delay Constraint
# set_input_delay -clock $input_clock -max [expr $input_clock_period/2 - $dv_bfe] [get_ports $input_ports];
# set_input_delay -clock $input_clock -min $dv_are                                [get_ports $input_ports];
# set_input_delay -clock $input_clock -max [expr $input_clock_period/2 - $dv_bre] [get_ports $input_ports] -clock_fall -add_delay;
# set_input_delay -clock $input_clock -min $dv_afe                                [get_ports $input_ports] -clock_fall -add_delay;
```

If fmc_input_clk_period   is 2 then dv_are + dv_bfe <= fmc_input_clk_period  / 2 correct?

Reply:
Yes. In practice to run above 500MHz you can set dv_are + dv_bfe > fmc_input_clk_period / 2, since Vivado is super pessimistic in timing. Strictly follow dv_are + dv_bfe <= fmc_input_clk_period / 2 correct may fail in timing.